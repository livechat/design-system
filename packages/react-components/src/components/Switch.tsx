import * as React from 'react';
import cx from 'classnames';
import classNames from 'classnames/bind';

import { noop } from './constants';

const baseClass = 'switch';

enum SwitchSize {
  Basic = 'basic',
  Compact = 'compact',
}

enum SwitchStyles {
  Enabled = 'enabled',
  Disabled = 'disabled',
}

export interface IProps {
  className?: string;
  defaultOn?: boolean;
  disabled?: boolean;
  innerRef?: React.LegacyRef<HTMLInputElement> | undefined;
  name?: string;
  on?: boolean;
  onChange?(e: React.FormEvent, value: boolean): void;
  size?: SwitchSize;
}

export const Switch: React.FC<IProps> = ({
  className = '',
  defaultOn = false,
  disabled = false,
  name = baseClass,
  on,
  onChange = noop,
  size = SwitchSize.Basic,
  innerRef,
  ...props
}) => {
  const getEnabledValue = () => {
    return on !== undefined ? on : defaultOn;
  };
  const [enabled, setEnabled] = React.useState(getEnabledValue());

  React.useEffect(() => {
    if (on !== undefined) {
      setEnabled(on);
    }
  }, [on]);

  const valueStyles = enabled ? SwitchStyles.Enabled : SwitchStyles.Disabled;
  const mergedClassNames = cx(baseClass, [`${baseClass}--${size}`], className);

  const handleChange = (e: React.FormEvent) => {
    const hasOnChangePassed = onChange !== noop;
    if (hasOnChangePassed) {
      onChange(e, enabled);
      return;
    }
    e.stopPropagation();
    setEnabled((prevEnabled) => !prevEnabled);
  };

  return (
    <span className={mergedClassNames}>
      <input
        type="checkbox"
        className={`${baseClass}__input`}
        onChange={handleChange}
        checked={enabled}
        name={name}
        ref={innerRef}
        disabled={disabled}
        test-id="foo"
        {...props}
      />
      <span className={`${baseClass}__container`}>
        <span
          className={classNames(
            `${baseClass}__track`,
            `${baseClass}__track--${valueStyles}`
          )}
        />
        <span
          className={classNames(
            `${baseClass}__slider`,
            `${baseClass}__slider--${size}`,
            `${baseClass}__slider--${size}--${valueStyles}`
          )}
        />
      </span>
    </span>
  );
};
