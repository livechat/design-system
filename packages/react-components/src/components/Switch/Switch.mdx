import { Meta, Title, Canvas, ArgTypes } from '@storybook/addon-docs/blocks';

import * as Stories from './Switch.stories';

<Meta of={Stories} />

<Title>Switch</Title>

[Intro](#Intro) | [Component API](#ComponentAPI) | [Content Spec](#ContentSpec)

## Intro <a id="Intro" />

<Canvas of={Stories.Default} sourceState="none" />

#### Example implementation

```jsx
<Switch on={false} state="regular" onChange={() => {}} />
```

## Component API <a id="ComponentAPI" />

<ArgTypes of={Stories.Default} sort="requiredFirst" />

## Content Spec <a id="ContentSpec" />

<a
  className="sb-unstyled"
  href="https://www.figma.com/file/9FDwjR8lYvincseDkKypC4/Component-Documentations?type=design&node-id=702-51275&mode=dev"
  target="_blank"
>
  Go to Figma documentation
</a>
