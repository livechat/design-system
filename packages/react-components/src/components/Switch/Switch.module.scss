$switch-transition-duration: 300ms;
$switch-transition-timing-function: cubic-bezier(0.33, 0, 0.67, 1);

$base-class: 'switch';
$large-height: 24px;
$large-width: 44px;
$large-size: $large-height - 4px;
$large-move: $large-width - $large-size;

$compact-height: 16px;
$compact-width: 32px;
$compact-size: $compact-height - 4px;
$compact-move: $compact-width - $compact-size;

$medium-height: 20px;
$medium-width: 36px;
$medium-size: $medium-height - 4px;
$medium-move: $medium-width - $medium-size;

/* stylelint-disable max-nesting-depth */
.#{$base-class} {
  display: inline-block;
  position: relative;
  outline: none;

  &:hover {
    $hovered-track: #{$base-class}__track;

    .#{$hovered-track}--enabled.#{$hovered-track}--on {
      background-color: var(--color-positive-hover);
    }

    .#{$hovered-track}--enabled.#{$hovered-track}--off {
      background-color: var(--surface-secondary-hover);
    }
  }

  &--large {
    width: $large-width;
    height: $large-height;
  }

  &--compact {
    width: $compact-width;
    height: $compact-height;
  }

  &--medium {
    width: $medium-width;
    height: $medium-height;
  }

  &__input {
    opacity: 0;
    margin: 0;
    cursor: pointer;
    width: 100%;
    height: 100%;

    &--disabled {
      cursor: not-allowed;
    }
  }

  &__input:focus + .#{$base-class}__container {
    .#{$base-class}__track {
      box-shadow: 0 0 1px 2px rgb(var(--color-action-default-rgb) 0.7);
    }
  }

  &__container {
    display: flex;
    position: absolute;
    top: 0;
    left: 0;
    align-items: center;
    width: 100%;
    height: 100%;
    pointer-events: none;
  }

  &__track {
    flex: 1;
    transition-duration: $switch-transition-duration;
    transition-property: background-color;
    transition-timing-function: $switch-transition-timing-function;
    border-radius: 24px;
    height: 100%;

    &--on#{&}--enabled {
      background-color: var(--color-positive-default);
    }

    &--on#{&}--disabled {
      background-color: var(--color-positive-disabled);
    }

    &--off#{&}--enabled {
      background-color: var(--surface-secondary-default);
    }

    &--off#{&}--disabled {
      background-color: var(--surface-secondary-disabled);
    }
  }

  &__slider {
    display: flex;
    position: absolute;
    top: 50%;
    align-items: center;
    justify-content: center;
    transform: translateY(-50%);
    transition-duration: $switch-transition-duration;
    transition-property: left, right;
    transition-timing-function: $switch-transition-timing-function;
    border-radius: 50%;
    background: var(--content-white-locked);

    &--large {
      width: calc(#{$large-size});
      height: calc(#{$large-size});

      &--on {
        right: 3px;
        left: calc(#{$large-move} - 3px);
      }

      &--off {
        right: calc(#{$large-move} - 3px);
        left: 3px;
      }
    }

    &--compact {
      width: calc(#{$compact-size});
      height: calc(#{$compact-size});

      &--on {
        right: 2px;
        left: calc(#{$compact-move} - 2px);
      }

      &--off {
        right: calc(#{$compact-move} - 2px);
        left: 2px;
      }
    }

    &--medium {
      width: calc(#{$medium-size});
      height: calc(#{$medium-size});

      &--on {
        right: 3px;
        left: calc(#{$medium-move} - 3px);
      }

      &--off {
        right: calc(#{$medium-move} - 3px);
        left: 3px;
      }
    }
  }

  &__loader {
    width: 100%;
    height: 100%;

    &--compact {
      padding: 1.5px;
    }

    &--medium {
      padding: 2px;
    }

    &--large {
      padding: 3px;
    }

    > div {
      width: 100%;
      height: 100%;
    }

    div:last-child {
      border-width: 2px;
    }
  }

  &__icon {
    color: var(--content-black-locked);
  }
}
