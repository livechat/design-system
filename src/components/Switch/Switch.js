import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';

const acceptedSizes = ['basic', 'compact'];
const baseClass = 'switch';
const cx = classNames.bind(styles);
const noop = () => {};

class SwitchComponent extends React.PureComponent {
  state = {
    enabled: this.isControlledByProps() ? this.props.on : this.props.defaultOn,
    prevPropsOn: this.props.on // eslint-disable-line react/no-unused-state
  };

  static getDerivedStateFromProps(props, state) {
    if (props.on !== state.prevPropsOn) {
      return {
        enabled: props.on,
        prevPropsOn: props.on
      };
    }
    return null;
  }

  handleChange = e => {
    const hasCb = this.props.onChange !== noop;
    if (hasCb) {
      this.props.onChange(e, this.state.enabled);
      return;
    }
    e.stopPropagation();
    this.setState(prevState => ({
      enabled: !prevState.enabled
    }));
  };

  isControlledByProps() {
    return this.props.on !== undefined;
  }

  render() {
    const {
      className,
      defaultOn,
      onChange,
      innerRef,
      on,
      size,
      name,
      ...restProps
    } = this.props;

    const { enabled } = this.state;
    const mergedClassNames = getMergedClassNames(
      cx({
        [baseClass]: true,
        [`${baseClass}--${size}`]: acceptedSizes.some(s => s === size)
      }),
      className
    );
    const valueStyles = enabled ? 'enabled' : 'disabled';

    return (
      <span className={mergedClassNames}>
        <input
          type="checkbox"
          className={styles[`${baseClass}__input`]}
          onChange={this.handleChange}
          checked={enabled}
          name={name}
          ref={innerRef}
          {...restProps}
        />
        <span className={styles[`${baseClass}__container`]}>
          <span
            className={classNames(
              styles[`${baseClass}__track`],
              styles[`${baseClass}__track--${valueStyles}`]
            )}
          />
          <span
            className={classNames(
              styles[`${baseClass}__slider`],
              styles[`${baseClass}__slider--${size}`],
              styles[`${baseClass}__slider--${size}--${valueStyles}`]
            )}
          />
        </span>
      </span>
    );
  }
}

const basePropTypes = {
  className: PropTypes.string,
  defaultOn: PropTypes.bool,
  /**
   * HTML disabled property
   */
  disabled: PropTypes.bool,
  name: PropTypes.string,
  on: PropTypes.bool,
  onChange: PropTypes.func,
  size: PropTypes.oneOf(acceptedSizes)
};

/* eslint-disable react/default-props-match-prop-types */
const baseDefaultProps = {
  defaultOn: false,
  onChange: noop,
  size: 'basic',
  name: baseClass
};

SwitchComponent.propTypes = {
  ...basePropTypes,
  innerRef: PropTypes.instanceOf(
    typeof Element === 'undefined' ? () => {} : Element
  )
};

SwitchComponent.defaultProps = baseDefaultProps;

const Switch = React.forwardRef((props, ref) => (
  <SwitchComponent innerRef={ref} {...props} />
));

Switch.propTypes = basePropTypes;

Switch.defaultProps = baseDefaultProps;

export default Switch;
