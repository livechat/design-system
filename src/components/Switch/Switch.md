Use toggles to immediately switch between two states (e.g. ON / OFF)

Basic size
```js
initialState = {
  isOn: false
};
const onChange = (e) => setState({isOn: !state.isOn});

<div style={{display: 'flex', alignItems: 'center'}}>
  <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
  <Switch onChange={onChange} on={state.isOn} />
</div>
```

```js noeditor
initialState = {
  isOn: true
};
const onChange = (e) => setState({isOn: !state.isOn});

  <ComponentHtmlMarkup>
    <div style={{display: 'flex', alignItems: 'center'}}>
      <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
      <Switch onChange={onChange} on={state.isOn} />
    </div>
  </ComponentHtmlMarkup>
```

Compact size
```js
initialState = {
  isOn: true
};
const onChange = (e) => setState({isOn: !state.isOn});

<div style={{display: 'flex', alignItems: 'center'}}>
  <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
  <Switch size="compact" onChange={onChange} on={state.isOn}/>
</div>
```
```js noeditor
initialState = {
  isOn: true
};
const onChange = (e) => setState({isOn: !state.isOn});
<ComponentHtmlMarkup>
  <div style={{display: 'flex', alignItems: 'center'}}>
    <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
    <Switch size="compact" onChange={onChange} on={state.isOn}/>
  </div>
</ComponentHtmlMarkup>
```

Based on component state
```js
  <div style={{display: 'flex', alignItems: 'center'}}>
    <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
    <Switch size="compact" defaultOn/>
  </div>
```

```js noeditor
<ComponentHtmlMarkup>
  <div style={{display: 'flex', alignItems: 'center'}}>
    <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
    <Switch size="compact" defaultOn/>
  </div>
</ComponentHtmlMarkup>
```

<h3>Best practices</h3>
<ul>
  <li>Donâ€™t require users to press a button to confirm the change. </li>
  <li>If you need users to press a button, use checkboxes instead. </li>
  <li>Write clear labels to make the choice less ambigous.</li>
  <li>Use positive and active wording for labels </li>
</ul>