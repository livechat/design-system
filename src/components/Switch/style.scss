@import "../../scss/variables/colors";
@import "../../scss/variables/transitions";

$base-class: "switch";
$basicHeight: 28px;
$basicWidth: 44px;
$basicSize: $basicHeight - 4px;
$basicMove: $basicWidth - $basicSize;

$compactHeight: 16px;
$compactWidth: 25px;
$compactSize: $compactHeight - 2px;
$compactMove: $compactWidth - $compactSize;

.#{$base-class} {
  position: relative;
  display: inline-block;
  outline: none;

  &:hover {
    .#{$base-class}__track {
      &--enabled {
        background-color: darken($success-color, 10);
      }

      &--disabled {
        background-color: darken($field-background-color, 10);
      }
    }
  }

  &--basic {
    width: $basicWidth;
    height: $basicHeight;
  }

  &--compact {
    width: $compactWidth;
    height: $compactHeight;
  }

  &__input {
    width: 100%;
    height: 100%;
    margin: 0;
    opacity: 0;
    cursor: pointer;
  }

  &__input:focus + .#{$base-class}__container {
    .#{$base-class}__track {
      box-shadow: 0 0 1px 2px rgba($hover-primary-color, 0.5);
    }
  }

  &__container {
    position: absolute;
    top: 0;
    left: 0;
    display: flex;
    align-items: center;
    width: 100%;
    height: 100%;
    pointer-events: none;
  }

  &__track {
    flex: 1;
    height: 100%;
    border-radius: 24px;
    transition-property: background-color;
    transition-duration: $switch-transition-duration;
    transition-timing-function: $switch-transition-timing-function;

    &--enabled {
      background-color: $success-color;
    }

    &--disabled {
      background-color: $field-background-color;
    }
  }

  &__slider {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    border-radius: 50%;
    background: white;
    transition-property: left, right;
    transition-duration: $switch-transition-duration;
    transition-timing-function: $switch-transition-timing-function;

    &--basic {
      width: calc(#{$basicSize});
      height: calc(#{$basicSize});

      &--enabled {
        left: calc(#{$basicMove} - 2px);
        right: 2px;
      }

      &--disabled {
        left: 2px;
        right: calc(#{$basicMove} - 2px);
      }
    }

    &--compact {
      width: calc(#{$compactSize});
      height: calc(#{$compactSize});

      &--enabled {
        left: calc(#{$compactMove} - 2px);
        right: 2px;
      }

      &--disabled {
        left: 2px;
        right: calc(#{$compactMove} - 2px);
      }
    }
  }
}
