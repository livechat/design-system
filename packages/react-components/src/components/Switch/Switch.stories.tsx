import * as React from 'react';
import { ComponentMeta, Story } from '@storybook/react';

import { StoryDescriptor } from '../../stories/components/StoryDescriptor';
import { DISABLED_CONTROLS } from '../../utils/story-parameters';

import { Switch, SwitchProps } from './Switch';

export default {
  title: 'Components/Switch',
  component: Switch,
} as ComponentMeta<typeof Switch>;

export const Default: Story<SwitchProps> = (args: SwitchProps) => (
  <Switch {...args} onChange={undefined} />
);
Default.storyName = 'Switch';

export const States: Story = (): JSX.Element => (
  <>
    <StoryDescriptor title="Regular">
      <Switch on={true} state="regular" />
      <Switch on={false} state="regular" />
    </StoryDescriptor>
    <StoryDescriptor title="Loading">
      <Switch on={true} state="loading" />
      <Switch on={false} state="loading" />
    </StoryDescriptor>
    <StoryDescriptor title="Locked">
      <Switch on={true} state="locked" />
      <Switch on={false} state="locked" />
    </StoryDescriptor>
  </>
);
States.parameters = DISABLED_CONTROLS;

export const Availability: Story = (): JSX.Element => (
  <>
    <StoryDescriptor title="Enabled">
      <Switch on={true} />
      <Switch on={false} />
    </StoryDescriptor>
    <StoryDescriptor title="Disabled">
      <Switch on={true} disabled={true} />
      <Switch on={false} disabled={true} />
    </StoryDescriptor>
  </>
);
Availability.parameters = DISABLED_CONTROLS;

export const Sizes: Story = (): JSX.Element => (
  <>
    <StoryDescriptor title="Compact">
      <Switch size="compact" />
    </StoryDescriptor>
    <StoryDescriptor title="Medium">
      <Switch size="medium" />
    </StoryDescriptor>
    <StoryDescriptor title="Large">
      <Switch size="large" />
    </StoryDescriptor>
  </>
);
Sizes.parameters = DISABLED_CONTROLS;
