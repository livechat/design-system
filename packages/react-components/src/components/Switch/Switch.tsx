import * as React from 'react';
import cx from 'clsx';
import { LockBlack as LockIcon } from '@livechat/design-system-icons/react/material';

import { Icon, IconSize } from '../../components/Icon';
import noop from '../../utils/noop';

import styles from './Switch.module.scss';
import { Loader } from '../Loader';

export const baseClass = 'switch';

export type SwitchSize = 'compact' | 'medium' | 'large';
export type SwitchState = 'regular' | 'loading' | 'locked';

export interface SwitchProps {
  className?: string;
  defaultOn?: boolean;
  disabled?: boolean;
  innerRef?: React.LegacyRef<HTMLInputElement> | undefined;
  name?: string;
  on?: boolean;
  onChange?(e: React.FormEvent, value: boolean): void;
  size?: SwitchSize;
  state?: SwitchState;
}

export const Switch: React.FC<SwitchProps> = ({
  className = '',
  defaultOn = false,
  disabled = false,
  name = baseClass,
  on,
  onChange = noop,
  size = 'large',
  state = 'regular',
  innerRef,
  ...props
}) => {
  const [checked, setChecked] = React.useState(() =>
    on !== undefined ? on : defaultOn
  );

  React.useEffect(() => {
    if (on !== undefined) {
      setChecked(on);
    }
  }, [on]);

  const isLoading = state === 'loading';
  const isLocked = state === 'locked';
  const iconSize: IconSize = size === 'large' ? 'small' : 'xsmall';
  const toggleStyles = checked ? 'on' : 'off';
  const shouldBehaveAsDisabled = disabled || isLoading || isLocked;
  const availabilityStyles = shouldBehaveAsDisabled ? 'disabled' : 'enabled';
  const mergedClassNames = cx(
    styles[baseClass],
    styles[`${baseClass}--${size}`],
    className
  );

  const handleChange = (e: React.FormEvent) => {
    const hasOnChangePassed = onChange !== noop;
    if (hasOnChangePassed) {
      onChange(e, checked);
      return;
    }
    e.stopPropagation();
    setChecked((prevEnabled) => !prevEnabled);
  };

  return (
    <span className={mergedClassNames}>
      <input
        type="checkbox"
        className={cx(
          styles[`${baseClass}__input`],
          styles[`${baseClass}__input--${availabilityStyles}`]
        )}
        onChange={handleChange}
        checked={checked}
        name={name}
        ref={innerRef}
        disabled={shouldBehaveAsDisabled}
        test-id="foo"
        {...props}
      />
      <span className={styles[`${baseClass}__container`]}>
        <span
          className={cx(
            styles[`${baseClass}__track`],
            styles[`${baseClass}__track--${toggleStyles}`],
            styles[`${baseClass}__track--${availabilityStyles}`]
          )}
        />
        <span
          className={cx(
            styles[`${baseClass}__slider`],
            styles[`${baseClass}__slider--${size}`],
            styles[`${baseClass}__slider--${size}--${toggleStyles}`]
          )}
        >
          {isLoading && (
            <Loader
              className={cx(
                styles[`${baseClass}__loader`],
                styles[`${baseClass}__loader--${size}`]
              )}
            />
          )}
          {isLocked && (
            <Icon
              className={styles[`${baseClass}__icon`]}
              data-testid="lock-icon"
              size={iconSize}
              source={LockIcon}
              kind="primary"
            />
          )}
        </span>
      </span>
    </span>
  );
};
