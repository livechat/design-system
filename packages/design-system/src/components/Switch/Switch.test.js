import * as React from 'react';
import { shallow } from 'enzyme';
import Switch from './Switch';

describe('Switch', () => {
  const toggleFn = jest.fn();

  it('should render basic Switch', () => {
    const component = shallow(<Switch onChange={toggleFn} />);

    expect(component).toMatchSnapshot();
  });

  it('should call on toggle when clicked', () => {
    const wrapper = shallow(<Switch onChange={toggleFn} />).dive();
    const checkbox = wrapper.find('input[type="checkbox"]');
    checkbox.simulate('change');
    expect(toggleFn).toHaveBeenCalled();
  });
});
