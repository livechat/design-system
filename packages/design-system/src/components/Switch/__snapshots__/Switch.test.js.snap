// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render basic Switch 1`] = `
<SwitchComponent
  defaultOn={false}
  innerRef={null}
  name="switch"
  onChange={[MockFunction]}
  size="basic"
/>
`;
